test/unit_test/simulator.cc

```diff
@@ -0,0 +1,127 @@
+#include "simulator.h"
+
+namespace simulator
+{
+cb_sim::cb_sim(int seed) : seed(seed) { srand(seed); }
+
+float cb_sim::get_cost(std::map<std::string, std::string> context, std::string action)
+{
+  if (context["user"] == "Tom")
+  {
+    if (context["time_of_day"] == "morning" && action == "politics") { return USER_LIKED_ARTICLE; }
+    else if (context["time_of_day"] == "afternoon" && action == "music")
+    {
+      return USER_LIKED_ARTICLE;
+    }
+  }
+  else if (context["user"] == "Anna")
+  {
+    if (context["time_of_day"] == "morning" && action == "sports") { return USER_LIKED_ARTICLE; }
+    else if (context["time_of_day"] == "afternoon" && action == "politics")
+    {
+      return USER_LIKED_ARTICLE;
+    }
+  }
+  return USER_DISLIKED_ARTICLE;
+}
+// todo: skip text format and create vw example directly
+std::vector<std::string> cb_sim::to_vw_example_format(
+    std::map<std::string, std::string> context, std::string chosen_action, float cost, float prob)
+{
+  std::vector<std::string> multi_ex_str;
+  multi_ex_str.push_back(fmt::format("shared |User user={} time_of_day={}", context["user"], context["time_of_day"]));
+  for (const auto& action : actions)
+  {
+    std::ostringstream ex;
+    if (action == chosen_action) { ex << fmt::format("0:{}:{} ", cost, prob); }
+    ex << fmt::format("|Action article={}", action);
+    multi_ex_str.push_back(ex.str());
+  }
+  return multi_ex_str;
+}
+
+std::pair<int, float> cb_sim::sample_custom_pmf(std::vector<float> pmf)
```

Comment:
```suggestion
std::pair<int, float> cb_sim::sample_custom_pmf(const std::vector<float>& pmf)
```

Reply:
This actually changes the values of pmf so it can't be `const`. Let me know if all the other changes look good